cryptol-remote-api/python/cryptol/commands.py

```diff
@@ -161,14 +183,62 @@ def __init__(self, connection : HasProtocolState, expr : Any) -> None:
     def process_result(self, res : Any) -> Any:
         return res['type schema']
 
+
 class SmtQueryType(str, Enum):
     PROVE = 'prove'
     SAFE  = 'safe'
     SAT   = 'sat'
 
-class CryptolProveSat(argo.Command):
+class SmtResult(ABC):
```

Comment:
I'm wondering if it would be clearer for more users (me included) if `sat`, `prove`, and `safe` returned types more specialized for the question they are answering, similar to what the Cryptol REPL reports. E.g.,

`:prove` at the REPL reports counterexample or QED:
```
Cryptol> :prove 1+1 == 0
Counterexample
(1 + 1 == 0) = False
(Total Elapsed Time: 0.009s, using "Z3")

Cryptol> :prove 1+1 == 2
Q.E.D.
(Total Elapsed Time: 0.009s, using "Z3")
```

so maybe `prove` in python returns `Union[CounterExample, QED]`?

`:sat` simply tells you whether or not the expression is satisfiable:
```
Cryptol> :sat \(x:[8]) -> 1+1 == x
Satisfiable
(\(x : [8]) -> 1 + 1 == x) 0x02 = True
(Total Elapsed Time: 0.010s, using "Z3")

Cryptol> :sat \(x:[8]) -> 1+x == 2+x
Unsatisfiable
(Total Elapsed Time: 0.010s, using "Z3")
Cryptol> 
```

perhaps `sat` in python returns `Union[Satisfiable, Unsatisfiable]`?

`:safe` reports either that indeed it was safe or it gives you a specific counter example:
```
Cryptol> :safe \(x:[8]) -> 1+x
Safe
(Total Elapsed Time: 0.009s, using "Z3")

Cryptol> :safe \(x:[8]) -> 1/x
Counterexample
(\(x : [8]) -> 1 / x) 0x00 ~> ERROR
division by 0
-- Backtrace --
(Cryptol::/) called at <interactive>:7:19--7:22
<interactive>::it called at <interactive>:7:7--7:22
(Total Elapsed Time: 0.011s, using "Z3")
```

perhaps `safe` in python returns `Union[Satisfiable, Unsatisfiable]`?


I get that we're sort of doing that with the `__bool__` on `SmtUnsatResult`, but then I immediately start to wonder why `SmtSatResult` doesn't have one, what that means, and I start getting confused and feel like I need run to a whiteboard real quick and whip out some greek notation to remind myself the model-theory implications of these various terms given the question we're posing... >_<

Reply:
Also, IMHO this might be one of those cases where we're better off _not_ using class-based inheritance, since we're talking about a small finite set of possibilities and we don't get any mileage out of polymorphic OO dispatch or the like. I.e., if the main win from inheritance is they all get a `qtype` field or a particular method, then maybe we should instead use a [Protocol](https://www.python.org/dev/peps/pep-0544/) for structural subtyping like we do [here](https://github.com/GaloisInc/argo/blob/37ce3b28832ff1fad0bf18630aeed59777a36fd4/python/argo_client/interaction.py#L10-L15) in `argo-client`.

Reply:
Thinking about this a little more and chatting with some folks internally about python idioms, I think I'm personally biased towards having more specific classes (i.e., not using class names like Satisfied/Unsatisfied unless that is specifically what the user was asking for in the request, otherwise it feels like a leaking of the SMT abstraction more than is ideal IMHO) and `__bool__` seems like a nice way to make testing simple (as was already nicely implemented in this PR)

Reply:
Check out the top of `synchronous.py` - there are now `Qed`, `Safe`, `Counterexample`, `Satisfiable`, and `Unsatisfiable` classes! `Qed`, `Safe`, and `Satisfiable` have `__bool__` methods which return `True`, `Counterexample` and `Unsatisfiable` have `__bool__` methods which return false, and `OfflineSmtQuery` has a bool method which throws an error (in `solver.py`).